numpy/core/function_base.py

```diff
@@ -350,9 +395,25 @@ def geomspace(start, stop, num=50, endpoint=True, dtype=None):
         start = start + 0j
         stop = stop + 0j
 
-    log_start = _nx.log10(start)
-    log_stop = _nx.log10(stop)
+    # Calculate the common ratio of the required geometric progression
+    if num == 1 or start == stop:
+        # Avoiding division by zero for `num = 1`
+        base = 10.0
+        log_start = _nx.log10(start)
+        log_stop = _nx.log10(stop)
+    else:
+        # Use Decimal type for real sequences
+        if start.imag == 0 and stop.imag == 0 and _nx.sign(start) == _nx.sign(stop):
+            base = _geom_common_ratio(start.real, stop.real, num, use_decimal=True)
+            base = float(base)
+        else:
+            base = _geom_common_ratio(start, stop, num)
+
+        log_base = _nx.log2(base)
+        log_start = _nx.log2(start) / log_base
+        log_stop = _nx.log2(stop) / log_base
```

Comment:
Special casing `base == 10` here would keep precision for the base 10 cases too,

Reply:
Might things be slightly more precise by using `_nx.log` above? Now, internally there are likely a number of divisions by `log2`

Reply:
The test
```
y = geomspace(1, 1e6, num=7)
array_equal(y, [1, 10, 100, 1e3, 1e4, 1e5, 1e6])
```
proves that `base == 10` works. What @endolith mentioned can only be fixed by special casing whenever `start` and `stop` are powers of 10, irrespective of the common ratio, but do we really need that?

Reply:
@mhvk That was what I tried in the beginning. For some reason then, the test `assert_array_equal(y, [1, 10, 100, 1e3, 1e4, 1e5, 1e6])` fails, because:
```python
>>> print(y)
[1.000000000000000e+00 9.999999999999998e+00 9.999999999999994e+01
 9.999999999999990e+02 9.999999999999989e+03 9.999999999999980e+04
 9.999999999999979e+05]
```

With `log2` we get exactly `[1.e+00 1.e+01 1.e+02 1.e+03 1.e+04 1.e+05 1.e+06]`. I don't know why honestly! Must be because `log` has an a irrational base.

Reply:
Using `log` mostly now. Special cases for `log2` and `log10` added.